# Exercise 6

## SVM & Regularization


For this homework we consider a set of observations on a number of red and white wine varieties involving their chemical properties and ranking by tasters. Wine industry shows a recent growth spurt as social drinking is on the rise. The price of wine depends on a rather abstract concept of wine appreciation by wine tasters, opinion among whom may have a high degree of variability. Pricing of wine depends on such a volatile factor to some extent. Another key factor in wine certification and quality assessment is physicochemical tests which are laboratory-based and takes into account factors like acidity, pH level, presence of sugar and other chemical properties. For the wine market, it would be of interest if human quality of tasting can be related to the chemical properties of wine so that certification and quality assessment and assurance process is more controlled.

Two datasets are available of which one dataset is on red wine and have 1599 different varieties and the other is on white wine and have 4898 varieties. All wines are produced in a particular area of Portugal. Data are collected on 12 different properties of the wines one of which is Quality, based on sensory data, and the rest are on chemical properties of the wines including density, acidity, alcohol content etc. All chemical properties of wines are continuous variables. Quality is an ordinal variable with possible ranking from 1 (worst) to 10 (best). Each variety of wine is tasted by three independent tasters and the final rank assigned is the median rank given by the tasters.

A predictive model developed on this data is expected to provide guidance to vineyards regarding quality and price expected on their produce without heavy reliance on volatility of wine tasters.

In [1]:
import pandas as pd
import numpy as np

In [2]:
data_r = pd.read_csv('https://github.com/albahnsen/PracticalMachineLearningClass/raw/master/datasets/Wine_data_red.csv')
data_w = pd.read_csv('https://github.com/albahnsen/PracticalMachineLearningClass/raw/master/datasets/Wine_data_white.csv')

In [3]:
data = data_w.assign(type = 'white')

data = data.append(data_r.assign(type = 'red'), ignore_index=True)
data.sample(5)

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
5953            8.2              0.64         0.27            2.00      0.095   
844             7.3              0.21         0.29            1.60      0.034   
2466            7.5              0.27         0.79           11.95      0.040   
1505            9.2              0.18         0.49            1.50      0.041   
4192            6.7              0.20         0.24            6.50      0.044   

      free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
5953                  5.0                  77.0  0.99747  3.13       0.62   
844                  29.0                 118.0  0.99170  3.30       0.50   
2466                 51.0                 159.0  0.99839  2.98       0.44   
1505                 39.0                 130.0  0.99450  3.04       0.49   
4192                 28.0                 100.0  0.99348  3.12       0.33   

      alcohol  quality   type  
5953      9.1        6    red  
844      11.0        8  white  
2466      8.7        5  white  
1505      9.8        7  white  
4192     10.2        6  white

# Exercise 6.1

Show the frecuency table of the quality by type of wine

In [6]:
t4 = pd.pivot_table(data, index = ['type'], values = 'quality', aggfunc = 'count')

t4

quality
type          
red       1599
white     4898

In [11]:
my_tab = pd.crosstab(index=data["type"], columns = data['quality'],margins = True)
my_tab

quality   3    4     5     6     7    8  9   All
type                                            
red      10   53   681   638   199   18  0  1599
white    20  163  1457  2198   880  175  5  4898
All      30  216  2138  2836  1079  193  5  6497

In [16]:
data['q_2'] = data.quality.map({3:0,4:0,5:0,6:1,7:1,8:1,9:1})
data.head()

#data['CH'] = data.Churn.map({'False.':0, 'True.':1})

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0            7.0              0.27         0.36            20.7      0.045   
1            6.3              0.30         0.34             1.6      0.049   
2            8.1              0.28         0.40             6.9      0.050   
3            7.2              0.23         0.32             8.5      0.058   
4            7.2              0.23         0.32             8.5      0.058   

   free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
0                 45.0                 170.0   1.0010  3.00       0.45   
1                 14.0                 132.0   0.9940  3.30       0.49   
2                 30.0                  97.0   0.9951  3.26       0.44   
3                 47.0                 186.0   0.9956  3.19       0.40   
4                 47.0                 186.0   0.9956  3.19       0.40   

   alcohol  quality   type  q_2  
0      8.8        6  white    1  
1      9.5        6  white    1  
2     10.1        6  white    1  
3      9.9        6  white    1  
4      9.9        6  white    1

In [18]:
datared = data[data['type'] == 'red']
datawhite = data[data['type'] == 'white']
datared.head()

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
4898            7.4              0.70         0.00             1.9      0.076   
4899            7.8              0.88         0.00             2.6      0.098   
4900            7.8              0.76         0.04             2.3      0.092   
4901           11.2              0.28         0.56             1.9      0.075   
4902            7.4              0.70         0.00             1.9      0.076   

      free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
4898                 11.0                  34.0   0.9978  3.51       0.56   
4899                 25.0                  67.0   0.9968  3.20       0.68   
4900                 15.0                  54.0   0.9970  3.26       0.65   
4901                 17.0                  60.0   0.9980  3.16       0.58   
4902                 11.0                  34.0   0.9978  3.51       0.56   

      alcohol  quality type  q_2  
4898      9.4        5  red    0  
4899      9.8        5  red    0  
4900      9.8        5  red    0  
4901      9.8        6  red    1  
4902      9.4        5  red    0

# SVM

# Exercise 6.2

* Standarized the features (not the quality)
* Create a binary target for each type of wine
* Create two Linear SVM's for the white and red wines, repectively.


# Exercise 6.3

Test the two SVM's using the different kernels (‘poly’, ‘rbf’, ‘sigmoid’)


# Exercise 6.4
Using the best SVM find the parameters that gives the best performance

'C': [0.1, 1, 10, 100, 1000], 'gamma': [0.01, 0.001, 0.0001]

# Exercise 6.5

Compare the results with other methods

# Regularization

# Exercise 6.6


* Train a linear regression to predict wine quality (Continous)

* Analyze the coefficients

* Evaluate the RMSE

# Exercise 6.7

* Estimate a ridge regression with alpha equals 0.1 and 1.
* Compare the coefficients with the linear regression
* Evaluate the RMSE

# Exercise 6.8

* Estimate a lasso regression with alpha equals 0.01, 0.1 and 1.
* Compare the coefficients with the linear regression
* Evaluate the RMSE

# Exercise 6.9

* Create a binary target

* Train a logistic regression to predict wine quality (binary)

* Analyze the coefficients

* Evaluate the f1score

# Exercise 6.10

* Estimate a regularized logistic regression using:
* C = 0.01, 0.1 & 1.0
* penalty = ['l1, 'l2']
* Compare the coefficients and the f1score